titus-server-master/src/main/java/com/netflix/titus/master/kubernetes/pod/legacy/DefaultAggregatingContainerEnvFactory.java

```diff
@@ -55,18 +57,22 @@ public DefaultAggregatingContainerEnvFactory(TitusRuntime titusRuntime, Containe
     }
 
     @Override
-    public Map<String, String> buildContainerEnv(Job<?> job, Task task) {
+    public Pair<List<String>, Map<String, String>> buildContainerEnv(Job<?> job, Task task) {
+        List<String> systemEnvNames = new ArrayList<>();
         Map<String, String> env = new HashMap<>();
         for (ContainerEnvFactory factory : orderedFactoryList) {
-            Map<String, String> envMap = factory.buildContainerEnv(job, task);
+            Pair<List<String>, Map<String, String>> incomingContainerEnv = factory.buildContainerEnv(job, task);
+            Map<String, String> incomingEnv = incomingContainerEnv.getRight();
+            List<String> incomingSystemEnv = incomingContainerEnv.getLeft();
             // Tracking conflicting env var for any two given factories
             env.keySet()
                     .stream()
-                    .filter(envMap::containsKey)
+                    .filter(incomingEnv::containsKey)
                     .forEach(envVarName -> incrementConflictCounter(envVarName, job.getId(), job.getJobDescriptor().getApplicationName()));
-            env.putAll(envMap);
+            env.putAll(incomingEnv);
+            systemEnvNames.addAll(incomingSystemEnv);
```

Comment:
This will not work well with the conflicts, unless the assumption is that we always start with the user env, and system overrides come later.

Reply:
It will work in the sense that `systemEnvNames` will have a list of all the keys, *even* if the values are not what we expected.